virttest/utils_libvirt/libvirt_vmxml.py

```diff
@@ -65,3 +74,55 @@ def remove_vm_devices_by_type(vm, device_type):
 
     if vm_was_running:
         vm.start()
+
+
+def create_vm_device_by_type(dev_type, dev_dict):
+    """
+    Create device by device type
+
+    :param dev_type: device type
+    :param dev_dict: dict for device
+    :return: device object
+    """
+    dev_obj = None
+    if dev_type == 'disk':
+        dev_obj = disk.Disk()
+    elif dev_type == 'controller':
+        dev_obj = controller.Controller()
+    elif dev_type == 'interface':
+        dev_obj = interface.Interface()
+    elif dev_type == 'channel':
+        dev_obj = channel.Channel()
+    elif dev_type == 'vsock':
+        dev_obj = vsock.Vsock()
+    elif dev_type == 'watchdog':
+        dev_obj = watchdog.Watchdog()
+    elif dev_type == 'rng':
+        dev_obj = rng.Rng()
+    elif dev_type == 'hostdev':
+        dev_obj = hostdev.Hostdev()
+
+    dev_obj.setup_attrs(**dev_dict)
+
+    return dev_obj
+
+
+def get_or_create_vm_device_if_absent(vmxml, dev_type, dev_dict, index=0):
+    """
+     Get specified device ,create the device if it does not exist
+
+    :param vmxml: domain VMXML instance
+    :param dev_type: device type
+    :param dev_dict: dict to create device
+    :param index: device index
+    :return: device object
+    """
+
+    try:
+        return libvirt.get_vm_device(vmxml, dev_type, index=index)
```

Comment:
If rename, please use setup_attrs() to update this returned device.

Reply:
done